{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#default_exp builtins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-04-01T01:45:50.088651Z",
     "start_time": "2021-04-01T01:45:49.974689Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Converted 00__init__.ipynb.\n",
      "Converted 00_core.ipynb.\n",
      "Converted 01_hdict.ipynb.\n",
      "Converted 02_seqs.ipynb.\n",
      "Converted 03_utils.ipynb.\n",
      "Converted 04_builtins.ipynb.\n",
      "Converted index.ipynb.\n"
     ]
    }
   ],
   "source": [
    "#hide\n",
    "from nbdev.showdoc import *\n",
    "from nbdev.export import notebook2script\n",
    "notebook2script()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-04-01T01:40:56.572579Z",
     "start_time": "2021-04-01T01:40:56.440454Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 80K\r\n",
      "-rw-r--r-- 1 jupyter-ken jupyter-ken  17K Mar 31 21:40 builtins.py\r\n",
      "-rw-r--r-- 1 jupyter-ken jupyter-ken  12K Mar 31 21:40 core.py\r\n",
      "-rw-r--r-- 1 jupyter-ken jupyter-ken 5.3K Mar 31 21:40 hdict.py\r\n",
      "-rw-r--r-- 1 jupyter-ken jupyter-ken  17K Mar 31 21:40 __init__.py\r\n",
      "-rw-r--r-- 1 jupyter-ken jupyter-ken 2.1K Mar 31 21:40 _nbdev.py\r\n",
      "drwxr-xr-x 2 jupyter-ken jupyter-ken 4.0K Mar 31 21:40 \u001b[0m\u001b[01;34m__pycache__\u001b[0m/\r\n",
      "-rw-r--r-- 1 jupyter-ken jupyter-ken 1.1K Mar 31 21:40 seqs.py\r\n",
      "-rw-r--r-- 1 jupyter-ken jupyter-ken 6.8K Mar 31 21:40 utils.py\r\n"
     ]
    }
   ],
   "source": [
    "ls -lh dask_rnaseq/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-04-01T00:17:18.791783Z",
     "start_time": "2021-04-01T00:17:18.785693Z"
    }
   },
   "outputs": [],
   "source": [
    "#export\n",
    "import os\n",
    "import glob\n",
    "\n",
    "import gzip\n",
    "import numpy as N\n",
    "import pandas as PD\n",
    "\n",
    "from prefect import task, Flow, Parameter, unmapped, case\n",
    "\n",
    "from dask_rnaseq.core import FileInFileOut, HBox, ptoml\n",
    "from dask_rnaseq.utils import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-03-31T00:21:01.263144Z",
     "start_time": "2021-03-31T00:21:01.222323Z"
    }
   },
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Act on FASTQ"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Countreads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-03-31T00:22:08.334378Z",
     "start_time": "2021-03-31T00:22:08.323567Z"
    }
   },
   "outputs": [],
   "source": [
    "# export\n",
    "class CountReads(FileInFileOut):\n",
    "    \"\"\"\n",
    "    Inputs: \n",
    "        1. Fastq (p.r1)\n",
    "    Outputs:\n",
    "        1. count.txt (=>p.info.nreads)\n",
    "    \n",
    "    \"\"\"\n",
    "    nickname = 'countreads'\n",
    "    i_tmpls = dict(\n",
    "        r1 = '{r1}'\n",
    "    )\n",
    "    o_tmpls = dict(\n",
    "        o1 = '{r1}.count.txt'\n",
    "    )\n",
    "            \n",
    "    def _run(self):\n",
    "        cmd = f\"cat {self._inps['r1']} | wc -l > {self._outs['o1']}\"\n",
    "        subprocesscall(cmd)\n",
    "        \n",
    "    def _read_stats(self):\n",
    "        nlines = int(open(self.outs['o1']).read())\n",
    "        return dict(nreads = int(nlines/4))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Flexbar"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-03-31T00:22:09.732618Z",
     "start_time": "2021-03-31T00:22:09.711220Z"
    }
   },
   "outputs": [],
   "source": [
    "# export\n",
    "class Flexbar(FileInFileOut):\n",
    "    \"\"\"\n",
    "    Inputs: \n",
    "        1. Fastq (p.r1, [p.r2])\n",
    "    Outputs: \n",
    "        1. Fastq (=>p.r1, [p.r2])\n",
    "    \"\"\"\n",
    "    nickname = 'flexbar'\n",
    "    t_tmpls = dict(\n",
    "        pr1 ='{prefix}_1.fastq', \n",
    "        pr2 ='{prefix}_2.fastq',\n",
    "        r1  ='{prefix}.fastq'\n",
    "    )\n",
    "    o_tmpls = dict(\n",
    "        pr1 ='{prefix}_1.fastq.gz', \n",
    "        pr2 ='{prefix}_2.fastq.gz', \n",
    "        r1  ='{prefix}.fastq.gz',\n",
    "        log ='{prefix}.log'\n",
    "    )\n",
    "    \n",
    "    def __init__(self, preset='Nextera', **kw):\n",
    "        self.preset = preset        \n",
    "        super().__init__(tag=preset, **kw)\n",
    "                \n",
    "    def _define_files(self):\n",
    "        p = self.p\n",
    "        self.paired = paired = (p.r2 != '')\n",
    "        if paired:\n",
    "            ikeys = ['r1','r2']\n",
    "            tkeys = ['pr1','pr2']\n",
    "            okeys = ['pr1','pr2','log']\n",
    "        else:\n",
    "            ikeys = ['r1']\n",
    "            tkeys = ['r1']\n",
    "            okeys = ['r1','log']\n",
    "        self.inps = {k: self.i_tmpls[k].format(**p) for k in ikeys}\n",
    "        self.tmps = {k: self.t_tmpls[k].format(**p) for k in tkeys}            \n",
    "        self.outs = {k: self.o_tmpls[k].format(**p) for k in okeys}\n",
    "                    \n",
    "    def _read_stats(self):\n",
    "        lines = open(self.outs['log']).readlines()\n",
    "        dic = dict(\n",
    "            processed = int([x for x in lines if x.startswith('Processed reads')][0].split()[2]),\n",
    "            remaining = int([x for x in lines if x.startswith('Remaining reads')][0].split()[2]),\n",
    "        )\n",
    "        return dic    \n",
    "    \n",
    "    def _modify_param_specific(self):\n",
    "        # swap inputs\n",
    "        pa = self.param\n",
    "        o = self.outs\n",
    "        if self.paired:\n",
    "            pa.r1, pa.r2 = o['pr1'], o['pr2']\n",
    "        else:\n",
    "            pa.r1 = o['r1']\n",
    "            \n",
    "    def _run(self):\n",
    "        flexbar = self.pcfg.get('execpath','flexbar')\n",
    "        nt = self.get_nthreads()\n",
    "        i,o,t = self._inps, self._outs, self._tmps\n",
    "        if self.paired:\n",
    "            cmd = f\"{flexbar} -r {i['r1']} -p {i['r2']} -aa {self.preset} -ap ON -n {nt} -t {self._prefix}\"\n",
    "        else:\n",
    "            cmd = f\"{flexbar} -r {i['r1']} -aa {self.preset} -n {nt} -t {self._prefix}\" \n",
    "        subprocesscall(cmd) # have to redirect\n",
    "        # compress with pigz\n",
    "        self.compress(self._tmps)\n",
    "                "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": true
   },
   "source": [
    "## Flexbar tests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-03-31T00:22:11.223615Z",
     "start_time": "2021-03-31T00:22:11.208800Z"
    },
    "hidden": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "nfs = \"./tests\"\n",
      "ramdisk = \"./tests2\"\n",
      "odir = \"./tests/outputs\"\n",
      "idir = \"./tests\"\n",
      "nfshost = \"host1\"\n",
      "sshuser = \"user1\"\n",
      "workers = [ \"host1\",]\n",
      "force = true\n",
      "did = \"datasetid\"\n",
      "maxthreads = 8\n",
      "sid = \"S69_R1\"\n",
      "r1 = \"./tests/5d_GH146mtdT_PN_Lee_L1_S69_R1_001.fastq.gz\"\n",
      "r2 = \"./tests/5d_GH146mtdT_PN_Lee_L1_S69_R2_001.fastq.gz\"\n",
      "\n",
      "[analysis]\n",
      "execpath = \"nonexistent\"\n",
      "\n",
      "[flexbar]\n",
      "execpath = \"/mnt/p1_nvme4/envs/bio/bin/flexbar\"\n",
      "\n",
      "[pigz]\n",
      "execpath = \"/mnt/p1_nvme4/envs/bio/bin/pigz\"\n",
      "\n",
      "[analysis.tag]\n",
      "extra = \"param1\"\n",
      "\n"
     ]
    }
   ],
   "source": [
    "cfg = HBox.from_toml(\"\"\"\n",
    "nfs = './tests'\n",
    "ramdisk = './tests2'\n",
    "odir = './tests/outputs'\n",
    "idir = './tests'\n",
    "nfshost = 'host1'\n",
    "sshuser = 'user1'\n",
    "workers = ['host1']\n",
    "force = true\n",
    "did = 'datasetid'\n",
    "maxthreads = 8\n",
    "\n",
    "[analysis]\n",
    "execpath = 'nonexistent'\n",
    "\n",
    "[analysis.tag]\n",
    "extra = 'param1'\n",
    "\n",
    "[flexbar]\n",
    "execpath = '/mnt/p1_nvme4/envs/bio/bin/flexbar'\n",
    "[pigz]\n",
    "execpath = '/mnt/p1_nvme4/envs/bio/bin/pigz'\n",
    "\"\"\")\n",
    "\n",
    "fb = Flexbar(preset='Nextera', cfg=cfg)\n",
    "\n",
    "sample = dict(sid = 'S69_R1',\n",
    "              r1 = './tests/5d_GH146mtdT_PN_Lee_L1_S69_R1_001.fastq.gz',\n",
    "              r2 = './tests/5d_GH146mtdT_PN_Lee_L1_S69_R2_001.fastq.gz')\n",
    "fb.set_param(sample)\n",
    "ptoml(fb.p)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-03-31T00:23:40.209111Z",
     "start_time": "2021-03-31T00:23:40.188207Z"
    },
    "hidden": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "({'r1': './tests/5d_GH146mtdT_PN_Lee_L1_S69_R1_001.fastq.gz',\n",
       "  'r2': './tests/5d_GH146mtdT_PN_Lee_L1_S69_R2_001.fastq.gz'},\n",
       " {'pr1': './tests/outputs/datasetid/flexbar.Nextera/S69_R1_1.fastq.gz',\n",
       "  'pr2': './tests/outputs/datasetid/flexbar.Nextera/S69_R1_2.fastq.gz',\n",
       "  'log': './tests/outputs/datasetid/flexbar.Nextera/S69_R1.log'},\n",
       " {'pr1': './tests/outputs/datasetid/flexbar.Nextera/S69_R1_1.fastq',\n",
       "  'pr2': './tests/outputs/datasetid/flexbar.Nextera/S69_R1_2.fastq'})"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fb._prep_dstdir()\n",
    "fb._define_files()\n",
    "fb.inps, fb.outs, fb.tmps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-03-30T21:10:23.604446Z",
     "start_time": "2021-03-30T21:10:23.251741Z"
    },
    "hidden": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "********** NFS => RAMDISK *********************************\n",
      "./tests/5d_GH146mtdT_PN_Lee_L1_S69_R1_001.fastq.gz=>suginok-p1:./tests2/5d_GH146mtdT_PN_Lee_L1_S69_R1_001.fastq.gz\n",
      "***********************************************************\n",
      "********** NFS => RAMDISK *********************************\n",
      "./tests/5d_GH146mtdT_PN_Lee_L1_S69_R2_001.fastq.gz=>suginok-p1:./tests2/5d_GH146mtdT_PN_Lee_L1_S69_R2_001.fastq.gz\n",
      "***********************************************************\n",
      "SHELL CMD ----------------------------------\n",
      "/mnt/p1_nvme4/envs/bio/bin/flexbar -r ./tests2/5d_GH146mtdT_PN_Lee_L1_S69_R1_001.fastq.gz -p ./tests2/5d_GH146mtdT_PN_Lee_L1_S69_R2_001.fastq.gz -aa Nextera -ap ON -n 8 -t ./tests2/outputs/datasetid/flexbar.Nextera/S69_R1\n",
      "--------- ----------------------------------\n",
      "SHELL CMD ----------------------------------\n",
      "/mnt/p1_nvme4/envs/bio/bin/pigz -p 8 --fast ./tests2/outputs/datasetid/flexbar.Nextera/S69_R1_1.fastq\n",
      "--------- ----------------------------------\n",
      "SHELL CMD ----------------------------------\n",
      "/mnt/p1_nvme4/envs/bio/bin/pigz -p 8 --fast ./tests2/outputs/datasetid/flexbar.Nextera/S69_R1_2.fastq\n",
      "--------- ----------------------------------\n",
      "********** RAMDISK => NFS *********************************\n",
      "suginok-p1:./tests2/outputs/datasetid/flexbar.Nextera/S69_R1_1.fastq.gz=>./tests/outputs/datasetid/flexbar.Nextera/S69_R1_1.fastq.gz\n",
      "***********************************************************\n",
      "********** RAMDISK => NFS *********************************\n",
      "suginok-p1:./tests2/outputs/datasetid/flexbar.Nextera/S69_R1_2.fastq.gz=>./tests/outputs/datasetid/flexbar.Nextera/S69_R1_2.fastq.gz\n",
      "***********************************************************\n",
      "********** RAMDISK => NFS *********************************\n",
      "suginok-p1:./tests2/outputs/datasetid/flexbar.Nextera/S69_R1.log=>./tests/outputs/datasetid/flexbar.Nextera/S69_R1.log\n",
      "***********************************************************\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<HBox: {'sid': 'S69_R1', 'r1': './tests/outputs/datasetid/flexbar.Nextera/S69_R1_1.fastq.gz', 'r2': './tests/outputs/datasetid/flexbar.Nextera/S69_R1_2.fastq.gz', 'delete': [('suginok-p1', './tests2/5d_GH146mtdT_PN_Lee_L1_S69_R1_001.fastq.gz'), ('suginok-p1', './tests2/5d_GH146mtdT_PN_Lee_L1_S69_R2_001.fastq.gz'), ('suginok-p1', './tests2/outputs/datasetid/flexbar.Nextera/S69_R1_1.fastq.gz'), ('suginok-p1', './tests2/outputs/datasetid/flexbar.Nextera/S69_R1_2.fastq.gz'), ('suginok-p1', './tests2/outputs/datasetid/flexbar.Nextera/S69_R1.log')], 'flexbar': {'Nextera': {'dstdir': './tests/outputs/datasetid/flexbar.Nextera', 'prefix': './tests/outputs/datasetid/flexbar.Nextera/S69_R1', 'inputs': ['5d_GH146mtdT_PN_Lee_L1_S69_R1_001.fastq.gz', '5d_GH146mtdT_PN_Lee_L1_S69_R2_001.fastq.gz'], 'outputs': ['S69_R1_1.fastq.gz', 'S69_R1_2.fastq.gz', 'S69_R1.log'], 'temps': ['S69_R1_1.fastq', 'S69_R1_2.fastq']}}, 'info': {'flexbar': {'Nextera': {'processed': 18384, 'remaining': 18368}}}, 'history': 'flexbar.Nextera'}>"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fb.run(sample)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-03-30T21:10:37.168941Z",
     "start_time": "2021-03-30T21:10:37.135252Z"
    },
    "hidden": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[Nextera]\n",
      "dstdir = \"./tests/outputs/datasetid/flexbar.Nextera\"\n",
      "prefix = \"./tests/outputs/datasetid/flexbar.Nextera/S69_R1\"\n",
      "inputs = [ \"5d_GH146mtdT_PN_Lee_L1_S69_R1_001.fastq.gz\", \"5d_GH146mtdT_PN_Lee_L1_S69_R2_001.fastq.gz\",]\n",
      "outputs = [ \"S69_R1_1.fastq.gz\", \"S69_R1_2.fastq.gz\", \"S69_R1.log\",]\n",
      "temps = [ \"S69_R1_1.fastq\", \"S69_R1_2.fastq\",]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "ptoml(fb.param['flexbar'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-03-30T21:10:45.623505Z",
     "start_time": "2021-03-30T21:10:45.465492Z"
    },
    "hidden": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\r\n",
      "               ________          __              \r\n",
      "              / ____/ /__  _  __/ /_  ____ ______\r\n",
      "             / /_  / / _ \\| |/ / __ \\/ __ `/ ___/\r\n",
      "            / __/ / /  __/>  </ /_/ / /_/ / /    \r\n",
      "           /_/   /_/\\___/_/|_/_.___/\\__._/_/     \r\n",
      "\r\n",
      "Flexbar - flexible barcode and adapter removal, version 3.5.0\r\n",
      "Developed with SeqAn, the library for sequence analysis\r\n",
      "\r\n",
      "Available on github.com/seqan/flexbar\r\n",
      "\r\n",
      "\r\n",
      "Local time:            Tue Mar 30 17:10:23 2021\r\n",
      "\r\n",
      "Number of threads:     8\r\n",
      "Bundled fragments:     256\r\n",
      "\r\n",
      "Target name:           ./tests2/outputs/datasetid/flexbar.Nextera/S69_R1\r\n",
      "File type:             fastq\r\n",
      "Reads file:            ./tests2/5d_GH146mtdT_PN_Lee_L1_S69_R1_001.fastq.gz\r\n",
      "Reads file 2:          ./tests2/5d_GH146mtdT_PN_Lee_L1_S69_R2_001.fastq.gz   (paired run)\r\n",
      "Adapter preset:        Nextera\r\n",
      "\r\n",
      "max-uncalled:          0\r\n",
      "min-read-length:       18\r\n",
      "\r\n",
      "adapter-pair-overlap:  ON\r\n",
      "adapter-trim-end:      RIGHT\r\n",
      "adapter-min-overlap:   3\r\n",
      "adapter-min-poverlap:  40\r\n",
      "adapter-error-rate:    0.1\r\n",
      "adapter-match:         1\r\n",
      "adapter-mismatch:     -1\r\n",
      "adapter-gap:          -6\r\n",
      "\r\n",
      "Adapter:               Sequence:\r\n",
      "Nextera-TruSight       CTGTCTCTTATACACATCT\r\n",
      "\r\n",
      "\r\n",
      "Processing reads ...done.\r\n",
      "\r\n",
      "Elapsed time:  < 1 sec\r\n",
      "\r\n",
      "\r\n",
      "Min, max, mean and median overlap of paired reads: 40 / 99 / 67 / 66\r\n",
      "\r\n",
      "Adapter removal statistics\r\n",
      "==========================\r\n",
      "Adapter:            Overlap removal:    Full length:\r\n",
      "Nextera-TruSight    7452                5696\r\n",
      "\r\n",
      "Min, max, mean and median overlap: 1 / 20 / 16 / 19\r\n",
      "\r\n",
      "\r\n",
      "Output file statistics\r\n",
      "======================\r\n",
      "Read file:               ./tests2/outputs/datasetid/flexbar.Nextera/S69_R1_1.fastq\r\n",
      "  written reads          9184\r\n",
      "  short reads            0\r\n",
      "Read file 2:             ./tests2/outputs/datasetid/flexbar.Nextera/S69_R1_2.fastq\r\n",
      "  written reads          9184\r\n",
      "  short reads            0\r\n",
      "\r\n",
      "\r\n",
      "Filtering statistics\r\n",
      "====================\r\n",
      "Processed reads                   18384\r\n",
      "  skipped due to uncalled bases      16   (12 uncalled in 8 pairs)\r\n",
      "  short prior to adapter removal      0\r\n",
      "  finally skipped short reads         0\r\n",
      "  skipped paired single reads         0\r\n",
      "Discarded reads overall              16\r\n",
      "Remaining reads                   18368   (99%)\r\n",
      "\r\n",
      "Processed bases   1838400\r\n",
      "Remaining bases   1559160   (84% of input)\r\n",
      "\r\n",
      "\r\n",
      "Flexbar completed adapter removal.\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "cat ./tests2/outputs/datasetid/flexbar.Nextera/S69_R1.log"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Mapper"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Dart"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-03-31T21:00:50.464400Z",
     "start_time": "2021-03-31T21:00:50.427179Z"
    },
    "code_folding": [
     47,
     80
    ]
   },
   "outputs": [],
   "source": [
    "# export\n",
    "class Dart(FileInFileOut):\n",
    "    \"\"\"\n",
    "    Inputs: \n",
    "        1. fastq (p.r1, [p.r2])\n",
    "    Outputs: \n",
    "        1. aligned.sorted.bam (=>p.bam)\n",
    "        2. unmapped.fastq (=>p.r1, [p.r2])\n",
    "    \"\"\"\n",
    "    nickname = 'dart'\n",
    "    t_tmpls = dict(\n",
    "        obam = '{prefix}.bam', \n",
    "        abam = '{prefix}.aligned.bam',\n",
    "        ubam = '{prefix}.unmapped.bam',\n",
    "        fq1 =  '{prefix}.{indexname}.unmapped_1.fastq',\n",
    "        fq2 =  '{prefix}.{indexname}.unmapped_2.fastq',\n",
    "        fq =   '{prefix}.{indexname}.unmapped.fastq',\n",
    "    )\n",
    "    o_tmpls = dict(\n",
    "        sbam = '{prefix}.aligned.sorted.bam', \n",
    "        sbai = '{prefix}.aligned.sorted.bam.bai', \n",
    "        log =  '{prefix}.{indexname}.log',\n",
    "        fq1 =  '{prefix}.{indexname}.unmapped_1.fastq.gz',\n",
    "        fq2 =  '{prefix}.{indexname}.unmapped_2.fastq.gz',                   \n",
    "        fq =   '{prefix}.{indexname}.unmapped.fastq.gz'\n",
    "    )\n",
    "    \n",
    "    def __init__(self, indexname, **kw):\n",
    "        self.indexname = indexname\n",
    "        super().__init__(tag=indexname, **kw)\n",
    "        self.cfg.indexname = indexname\n",
    "        \n",
    "    def _define_files(self):\n",
    "        p = self.p\n",
    "        self.paired = paired = (p.r2 != '')\n",
    "        if paired:\n",
    "            ikeys = ['r1','r2']\n",
    "            tkeys = ['obam','abam','ubam','fq1','fq2']\n",
    "            okeys = ['sbam','sbai','log','fq1','fq2']\n",
    "        else:\n",
    "            ikeys = ['r1']\n",
    "            tkeys = ['obam','abam','ubam','fq']\n",
    "            okeys = ['sbam','sbai','log','fq']\n",
    "        self.inps = {k: self.i_tmpls[k].format(**p) for k in ikeys}\n",
    "        self.tmps = {k: self.t_tmpls[k].format(**p) for k in tkeys}            \n",
    "        self.outs = {k: self.o_tmpls[k].format(**p) for k in okeys}\n",
    "    \n",
    "    def _read_stats(self):\n",
    "        log = self.outs['log']\n",
    "        dic = {}\n",
    "        def _extract_num(line):\n",
    "            return [int(s) for s in line.split() if s.isdigit()]\n",
    "        for line in open(log,'rt'):\n",
    "            if line.startswith('All the '):\n",
    "                dic['total_reads'],dic['duration'] = _extract_num(line)\n",
    "            elif line.startswith('\\t# of total mapped reads'):\n",
    "                dic['mapped'] = _extract_num(line)[0]\n",
    "            elif line.startswith('\\t# of paired sequences'):\n",
    "                dic['paired'] = _extract_num(line)[0]\n",
    "            elif line.startswith('\\t# of unique mapped'):\n",
    "                dic['unique'] = _extract_num(line)[0]\n",
    "            elif line.startswith('\\t# of multiple mapped'):\n",
    "                dic['multimap'] = _extract_num(line)[0]\n",
    "            elif line.startswith('\\t# of unmapped reads'):\n",
    "                dic['unmapped'] = _extract_num(line)[0]\n",
    "            elif line.startswith('\\t# of splice junctions'):\n",
    "                dic['spliced'] = _extract_num(line)[0]\n",
    "        return dic \n",
    "    \n",
    "    def _modify_param_specific(self):\n",
    "        # swap fastq inputs\n",
    "        pa = self.param\n",
    "        o = self.outs\n",
    "        if self.paired:\n",
    "            pa.r1, pa.r2 = o['fq1'], o['fq2']\n",
    "        else:\n",
    "            pa.r1 = o['fq']\n",
    "        # bam\n",
    "        pa.bam = o['sbam'] # sorted bam\n",
    "        \n",
    "    def _run0(self):\n",
    "        p = self.p\n",
    "        pc = self.pcfg\n",
    "        i = self._inps\n",
    "        t = self._tmps\n",
    "        o = self._outs\n",
    "        paired = self.paired\n",
    "        # 1st step: bam\n",
    "        dart = self.pcfg.execpath\n",
    "        if p.force or any([not os.path.exists(t[x]) for x in ['obam','abam','ubam']]): \n",
    "            nt = self.get_nthreads()\n",
    "            extra = pc.get('extra','')\n",
    "            idx = pc.index_prefix\n",
    "            if paired:\n",
    "                cmd = f\"\"\"{dart} -i {idx} -f {i['r1']} -f2 {i['r2']} -bo {t['obam']} -t {nt} {extra} \\\n",
    "                          | tr '\\r' '\\n' | tail -n 9 > {o['log']}\"\"\"\n",
    "            else:\n",
    "                cmd = f\"\"\"{dart} -i {idx} -f {i['r1']}               -bo {t['obam']} -t {nt} {extra} \\\n",
    "                          | tr '\\r' '\\n' | tail -n 9 > {o['log']}\"\"\"\n",
    "            subprocesscall(cmd)\n",
    "        # 2nd step: bam => aligned.bam, unmapped.fastq\n",
    "        samtools = p['samtools.execpath']\n",
    "        nt = self.get_nthreads('samtools')\n",
    "        if p.force or any([not os.path.exists(o[x]) for x in ['sbam','sbai']]):\n",
    "            inp = t['obam']\n",
    "            if paired:\n",
    "                cmd1 = f\"{samtools} view -@ {nt} -f 2 {inp} -o {t['abam']}\"\n",
    "                cmd2 = f\"{samtools} view -@ {nt} -F 2 {inp} -o {t['ubam']}\"\n",
    "            else:\n",
    "                cmd1 = f\"{samtools} view -@ {nt} -F 4 {inp} -o {t['abam']}\"\n",
    "                cmd2 = f\"{samtools} view -@ {nt} -f 4 {inp} -o {t['ubam']}\"\n",
    "            subprocesscall(cmd1)\n",
    "            subprocesscall(cmd2)\n",
    "            os.unlink(inp)\n",
    "            # convert unmapped.bam to fastq\n",
    "            bedtools = p['bedtools.execpath']\n",
    "            if paired:\n",
    "                cmd = f\"{bedtools} bamtofastq -i {t['ubam']} -fq {t['fq1']} -fq2 {t['fq2']}\"\n",
    "            else:\n",
    "                cmd = f\"{bedtools} bamtofastq -i {t['ubam']} -fq {t['fq']}\"\n",
    "            subprocesscall(cmd)\n",
    "            tgts = [t['fq1'],t['fq2']] if paired else [t['fq']]\n",
    "            self.compress(tgts)            \n",
    "        # 3rd step: aligned.bam => aligned.sorted.bam\n",
    "        if p.force or any([not os.path.exists(o[x]) for x in ['sbam']]):\n",
    "            cmd = f\"{samtools} sort {t['abam']} -o {o['sbam']} -@ {nt}\"\n",
    "            subprocesscall(cmd)\n",
    "            os.unlink(t['abam'])\n",
    "        # 4th step: bam index\n",
    "        if p.force or (not os.path.exists(o['sbai'])):\n",
    "            cmd = f\"{samtools} index {o['sbam']}\"\n",
    "            subprocesscall(cmd)    \n",
    "\n",
    "    def _run(self):\n",
    "        p = self.p\n",
    "        pc = self.pcfg\n",
    "        i = self._inps\n",
    "        t = self._tmps\n",
    "        o = self._outs\n",
    "        paired = self.paired\n",
    "        dart = self.pcfg.execpath\n",
    "        samtools = p.samtools.execpath\n",
    "        bedtools = p.bedtools.execpath\n",
    "        pigz = p.pigz.execpath\n",
    "        nt = self.get_nthreads()\n",
    "        nts = self.get_nthreads('samtools')\n",
    "        ntp = self.get_nthreads('pigz')\n",
    "        extra = pc.get('extra','')\n",
    "        idx = pc.index_prefix\n",
    "        if paired:\n",
    "            cmd = f\"\"\"{dart} -i {idx} -f {i['r1']} -f2 {i['r2']} -bo {t['obam']} -t {nt} {extra} \\\n",
    "                          | tr '\\r' '\\n' | tail -n 9 > {o['log']}; \\\n",
    "                      {samtools} view -@ {nts} -f 2 {t['obam']} -o {t['abam']}; \\\n",
    "                      {samtools} view -@ {nts} -F 2 {t['obam']} -o {t['ubam']}; \\\n",
    "                      rm {t['obam']}; \\\n",
    "                      {bedtools} bamtofastq -i {t['ubam']} -fq {t['fq1']} -fq2 {t['fq2']}; \\\n",
    "                      {pigz} -p {ntp} --fast {t['fq1']}; \\\n",
    "                      {pigz} -p {ntp} --fast {t['fq2']}; \\\n",
    "                      {samtools} sort {t['abam']} -o {o['sbam']} -@ {nts}; \\\n",
    "                      rm {t['abam']}; \\\n",
    "                      {samtools} index {o['sbam']}; \n",
    "                   \"\"\"\n",
    "        else:\n",
    "            cmd = f\"\"\"{dart} -i {idx} -f {i['r1']}               -bo {t['obam']} -t {nt} {extra} \\\n",
    "                          | tr '\\r' '\\n' | tail -n 9 > {o['log']}; \\\n",
    "                      {samtools} view -@ {nts} -F 4 {t['obam']} -o {t['abam']}; \\\n",
    "                      {samtools} view -@ {nts} -f 4 {t['obam']} -o {t['ubam']}; \\\n",
    "                      rm {t['obam']}; \\\n",
    "                      {bedtools} bamtofastq -i {t['ubam']} -fq {t['fq']}; \\\n",
    "                      {pigz} -p {ntp} --fast {t['fq1']}; \\\n",
    "                      {samtools} sort {t['abam']} -o {o['sbam']} -@ {nts}; \\\n",
    "                      rm {t['abam']}; \\\n",
    "                      {samtools} index {o['sbam']}; \n",
    "                   \"\"\"\n",
    "        subprocesscall(cmd)            \n",
    "\n",
    "# dart: https://github.com/hsinnan75/DART\n",
    "@task\n",
    "def dart_scatter_index(cfg, indexname):\n",
    "    \"scatter dart indices, return modified config\"\n",
    "    if (cfg.nfs=='') or (cfg.ramdisk=='') or (cfg.nfs== cfg.ramdisk):\n",
    "        return cfg\n",
    "    idxpre = cfg[f'dart.{indexname}.index_prefix']\n",
    "    nfshost = cfg.nfshost\n",
    "    remotes = cfg.workers #[x for x in p0.workers if x!=nfshost]\n",
    "    files = glob.glob(idxpre+'*')\n",
    "    # copy index files to all workers (except nfs host)\n",
    "    for host in remotes:\n",
    "        srcdsts = [(x,x.replace(cfg.nfs,cfg.ramdisk)) for x in files]\n",
    "        remote_put_files(host,srcdsts,cfg.sshuser)\n",
    "    cfg1 = cfg.copy()\n",
    "    idxpre1 = idxpre.replace(cfg.nfs,cfg.ramdisk)\n",
    "    cfg1[f'dart.{indexname}.index_prefix'] = idxpre1    \n",
    "    return cfg1\n",
    "    \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Gene Expression Read Counts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## FeatureCounts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-04-01T00:15:40.139062Z",
     "start_time": "2021-04-01T00:15:40.128718Z"
    }
   },
   "outputs": [],
   "source": [
    "# export\n",
    "class FeatureCounts(FileInFileOut):\n",
    "    \"\"\"\n",
    "    Inputs:\n",
    "        1. bam (p.bam)\n",
    "    Outputs:\n",
    "        1. counts (=>p.fccnts)\n",
    "        2. readmap (=>p.fcmaps)\n",
    "    \"\"\"\n",
    "    nickname = 'featurecounts'\n",
    "    i_tmpls = dict(\n",
    "        bam = '{bam}'\n",
    "    )\n",
    "    t_tmpls = dict(\n",
    "        count = '{prefix}.featurecounts.txt',\n",
    "        reads = '{prefix}.aligned.sorted.bam.featureCounts', # destination of each reads        \n",
    "    )\n",
    "    o_tmpls = dict(\n",
    "        count = '{prefix}.featurecounts.txt.gz',\n",
    "        reads = '{prefix}.aligned.sorted.bam.featureCounts.gz',\n",
    "        log = '{prefix}.featurecounts.txt.summary',\n",
    "    )\n",
    "    \n",
    "    def _read_stats(self):\n",
    "        df = PD.read_table(self.outs['log'], comment='#', names=['field','value'], index_col=[0])\n",
    "        return dict(df['value'])\n",
    "    \n",
    "    def _modify_param_specific(self):\n",
    "        # swap fastq inputs\n",
    "        pa = self.param\n",
    "        o = self.outs\n",
    "        pa.fccnts = o['count']\n",
    "        pa.fcmaps = o['reads']\n",
    "            \n",
    "    def _run(self):\n",
    "        pc = self.pcfg\n",
    "        exe = pc.execpath\n",
    "        i = self._inps\n",
    "        t = self._tmps\n",
    "        cmd = f\"{exe} -R CORE -t exon -g gene_id --primary -a {pc['gtf']} -o {t['count']} {i['bam']}\"\n",
    "        subprocesscall(cmd)\n",
    "        self.compress(t)\n",
    "                        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# SmartSeq3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-04-01T01:37:08.299814Z",
     "start_time": "2021-04-01T01:37:08.266903Z"
    },
    "code_folding": [
     17
    ]
   },
   "outputs": [],
   "source": [
    "# export\n",
    "class SS3CountTags(FileInFileOut):\n",
    "    \"\"\"\n",
    "    Inputs:\n",
    "        1. fastq (p.r1, p.r2) # assumes paired\n",
    "    Outputs:\n",
    "        1. csv (contents to p.info.ss3counttags)\n",
    "    \"\"\"\n",
    "    nickname = 'ss3counttags'\n",
    "    o_tmpls = dict(\n",
    "        tab = '{prefix}-{history}-{tag}.txt'\n",
    "    )\n",
    "    def __init__(self, tagseq='ATTGCGCAATG', **kw):\n",
    "        self.tagseq = tagseq\n",
    "        self.revseq = revcomp(tagseq)\n",
    "        super().__init__(**kw)\n",
    "        \n",
    "    def _read_stats(self):\n",
    "        if hasattr(self, '_cntsdf'):\n",
    "            return dict(self._cntsdf.sum(axis=0))\n",
    "        df = PD.read_table(self.outs['txt'], index_col=[0])\n",
    "        return dict(df.sum(axis=0))\n",
    "        \n",
    "    def _run(self):\n",
    "        p = self.p\n",
    "        fastq = FASTQP(p.r1,p.r2) # read all at once\n",
    "        cnts = {k: {} for k in ['p1','p2','rp1','rp2']}\n",
    "        for l1,l2 in fastq:\n",
    "            s1,s2 = l1[1],l2[1]\n",
    "            for k,s,t in [('p1',s1,tag),('p2',s2,tag),('rp1',s1,rtag),('rp2',s2,rtag)]:\n",
    "                pos = s.find(t)\n",
    "                cnts[k][pos] = cnts[k].get(pos,0) + 1 # count frequency of the position\n",
    "        a = PD.DataFrame(cnts).fillna(0).sort_index()\n",
    "        f = self._outs['txt']\n",
    "        a.to_csv(f, header=True, index=True, sep='\\t')\n",
    "        os.chmod(f, 0o775)\n",
    "        self._cntsdf = a\n",
    "\n",
    "\n",
    "\n",
    "class SS3RemoveTags(FileInFileOut):\n",
    "    \"\"\"\n",
    "    Inputs:\n",
    "        1. fastq (p.r1, p.r2) # paried\n",
    "    Outputs:\n",
    "        1. fastq (=> p.r1, p.r2) # tag removed\n",
    "        \n",
    "    Remove TAG and associated adapter/UMI from 5' side of read1.\n",
    "    If TAG is in r1 and rcTAG is in r2, then also remove 3' side from read2\n",
    "    Swap original (long id) to simple id, also group by unique seq (uniq seq id)\n",
    "    Record UMI, TAG position, uniq seq id, org id: as pandas dataframe.\n",
    "        \n",
    "    \"\"\"\n",
    "    nickname = 'ss3removetags'\n",
    "    o_tmpls = dict(\n",
    "        r1 = '{prefix}_1.fastq.gz',\n",
    "        r2 = '{prefix}_2.fastq.gz',\n",
    "        info = '{prefix}-info.txt.gz',\n",
    "    )\n",
    "    \n",
    "    def __init__(self, tagseq='ATTGCGCAATG', **kw):\n",
    "        self.tagseq = tagseq\n",
    "        super().__init__(**kw)\n",
    "        \n",
    "    def _read_stats(self):\n",
    "        dic = dict(\n",
    "            infopath = self.outs['info'],\n",
    "            stats = self._stats if hasattr(self, '_stats') else None,\n",
    "        )\n",
    "        return dic\n",
    "    \n",
    "    def _modify_param_specific(self):\n",
    "        pa = self.param\n",
    "        o = self.outs\n",
    "        pa.r1 = o['r1']\n",
    "        pa.r2 = o['r2']\n",
    "        \n",
    "    def _run(self):\n",
    "        p = self.p\n",
    "        pc = self.pcfg\n",
    "        fastq = FASTQP(p.r1,p.r2)\n",
    "        tag = self.tagseq\n",
    "        rtag = revcomp(tag)\n",
    "        lines1 = [] # filtered read 1 \n",
    "        lines2 = [] # filtered read 2\n",
    "        seqs = {} # seq => uid\n",
    "        info = [] # rows (orgid, newid, uniqid, len, tagpos, umiseq)\n",
    "        n = len(tag)\n",
    "        nu = pc.get('umibp',8)\n",
    "        minlen = pc.get('minlen',30)\n",
    "        idx = -1\n",
    "        for cid, (l1,l2) in enumerate(fastq):\n",
    "            s1,s2 = l1[1][:-1],l2[1][:-1] # seq\n",
    "            q1,q2 = l1[3][:-1],l2[3][:-1] # quality\n",
    "            p1 = s1.find(tag)       \n",
    "            p2 = s2.find(tag)\n",
    "            umi = '' # default no umi\n",
    "            if p1>=0: # NPA cleaved 5' end or using seqsite from FP\n",
    "                st = p1 + n + nu\n",
    "                umi = s1[st-nu:st] # umi seq\n",
    "                s1 = s1[st:] # trim 5'\n",
    "                q1 = q1[st:]\n",
    "                rp2 = s2.find(rtag)\n",
    "                if rp2>=0: # if the pair overlaps...\n",
    "                    ed = rp2 - nu\n",
    "                    s2 = s2[:ed] # trim 3'\n",
    "                    q2 = q2[:ed]\n",
    "            elif p2>=0: # NPB cleaved 5' end\n",
    "                st = p2 + n + nu\n",
    "                umi = s2[st-nu:st]\n",
    "                s2 = s2[st:]\n",
    "                q2 = q2[st:]\n",
    "                rp1 = s1.find(rtag)\n",
    "                if rp1>=0:\n",
    "                    ed = rp1 - nu\n",
    "                    s1 = s1[:ed]\n",
    "                    q1 = q1[:ed]\n",
    "            # new id = cid\n",
    "            # uniq id? (s1,s2)=> newid of first occurence\n",
    "            ovl,tot,sl1,sl2 = detect_overlap(s1,s2)\n",
    "            if (len(s1)>=minlen) and (len(s2)>=minlen):\n",
    "                uid = seqs.setdefault((s1,s2),cid)\n",
    "                lines1.append(f'@{cid}\\n{s1}\\n+\\n{q1}\\n')\n",
    "                lines2.append(f'@{cid}\\n{s2}\\n+\\n{q2}\\n')\n",
    "                idx += 1 # count lines added\n",
    "            else:\n",
    "                uid = -1\n",
    "            row = (l1[0],cid,uid,idx,tot,ovl,p1,p2,umi)\n",
    "            info.append(row)\n",
    "        # stats\n",
    "        self._stats = {'org':len(info),'filtered':len(lines1)}\n",
    "        # write\n",
    "        # info\n",
    "        infodf = PD.DataFrame(info, columns=['oid','nid','uid','idx','len','ovl','p1','p2','umi'])\n",
    "        o = self._outs\n",
    "        infodf.to_csv(o['info'], header=True, index=True, sep='\\t', compression='gzip')\n",
    "        os.chmod(o['info'], 0o775)\n",
    "        # read1,read2\n",
    "        gzip.open(o['r1'],'wt').write(''.join(lines1))\n",
    "        gzip.open(o['r2'],'wt').write(''.join(lines2))\n",
    "        os.chmod(o['r1'],0o775)\n",
    "        os.chmod(o['r2'],0o775)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "miniconda4.8.3",
   "language": "python",
   "name": "mini483"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "calc(100% - 180px)",
    "left": "10px",
    "top": "150px",
    "width": "203.571px"
   },
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
